ubi_manifest/worker/tasks/depsolver/modulemd_depsolver.py

```diff
@@ -0,0 +1,159 @@
+        self._searched_modules: Set[str] = set()
+        # output set of resolved modulemd packages
+        self.modules: List[UbiUnit] = []
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self._executor.__exit__(*args, **kwargs)
+
+    def run(self):
+        """
+        Run depsolver for each moudular dependency - recursively resolve all of
+        its modular dependencies and add binary and debug dependencies to list.
+        """
+        for item in self._modular_items:
+            modulemds_criteria = _get_criteria_for_modules(item.modulelist)
+            for module in item.modulelist:
+                self._update_searched_modules(module)
+            modules = f_proxy(
+                self._executor.submit(
+                    search_modulemds, modulemds_criteria, item.in_pulp_repos
+                )
+            )
+            # recurrently resolve dependencies for found modules
+            self._depsolve_modules(modules)
+        return self
+
+    def _depsolve_modules(self, modules):
+        """
+        Update modulemd output set with latest versions of modules, then find
+        dependencies for the modules and depsolve them.
+        """
+        filtered_modules = _get_modulemd_output_set(modules)
+        self.modules.extend(filtered_modules)
+
+        for module in filtered_modules:
+            # If dependencies is None, skip it
+            if not module.dependencies:
+                continue
+            # Get all dependencies and if not already processed, depsolve them as well
+            for dependency in module.dependencies:
+                if self._not_already_searched(dependency):
+                    self._update_searched_modules(dependency)
+                    modulemds_criteria = _get_criteria_for_modules(
+                        [
+                            dependency,
+                        ]
+                    )
+                    new_modules = f_proxy(
+                        self._executor.submit(
+                            search_modulemds, modulemds_criteria, self._input_repos
+                        )
+                    )
+                    self._depsolve_modules(new_modules)
+
+    def _update_searched_modules(self, module):
+        self._searched_modules.add(f"{module.name}:{module.stream}")
+
+    def _not_already_searched(self, module):
+        """Returns True if the module has not yet been searched for"""
+        return not f"{module.name}:{module.stream}" in self._searched_modules
+
+    def export(self) -> Dict[str, Dict[str, List[UbiUnit]]]:
+        """Returns a dictionary of depsolved modules, rpms, debug_rpms"""
+        out = {}
+        modules_out = {}
+        for module in self.modules:
+            modules_out.setdefault(module.associate_source_repo_id, []).append(module)
+        out["modules"] = modules_out
+        return out
+
+
+def _keep_n_latest_modules(modules, n=1):
```

Comment:
Can you please move those 3 functions to the utils.py?
There are already similar functions, so I believe it makes sense to move those as well.

Reply:
Moved.